### ISSUE TITLE
False positive: ``unnecessary lambda`` with kwargs in the call

### ISSUE THREAD

**Issue by clemux** [pylint-dev/pylint#9148]
### Bug description
```python
# pylint: disable=missing-module-docstring,undefined-variable,missing-function-docstring
def f(a, d):
    print(lambda x: a(x, **d))   
```

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:3:2: W0108: Lambda may not be necessary (unnecessary-lambda)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 0.00/10, +10.00)

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.0
Python 3.11.4 (main, Jun  9 2023, 07:59:55) [GCC 12.3.0]
```

### OS / Environment

Distributor ID: Ubuntu
Description:    Ubuntu 23.04
Release:        23.04
Codename:       lunar

**Comment by clemux** [comment]
I stumbled upon this false positive for W0108 (unnecessary-lambda) in the tests of pandas, for example [this](https://github.com/pandas-dev/pandas/blob/b5b8be04b2a63fc02e89650a740779e718d7a99b/pandas/tests/window/test_base_indexer.py#L178)

I've looked into the implementation, and [this part](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/base/basic_checker.py#L560C15-L568) looks weird to me:

```python
        if call.keywords:
            # Look for additional keyword arguments that are not part
            # of the lambda's signature
            lambda_kwargs = {keyword.name for keyword in node.args.defaults}
            if len(lambda_kwargs) != len(call_site.keyword_arguments):
                # Different lengths, so probably not identical
                return
            if set(call_site.keyword_arguments).difference(lambda_kwargs):
                return
```

As far as I can tell, `lambda_kwargs` is always empty because of [this condition](https://github.com/pylint-dev/pylint/blob/main/pylint/checkers/base/basic_checker.py#L528) at the beginning of the check

```python
        if node.args.defaults:
            return
```

In the case of my example, `len(call_site.keyword_arguments)` is empty. I don't know why it is empty, or even why the condition is made on this instead of simply comparing `call.keywords` and `node.args.kwargs`, but it explains why the warning is triggered here.

**Comment by clemux** [comment]
I've put my PR back into draft because of this new false positive in pandas :(

```python
(lambda df1, df2, **kwargs: merge(df2, df1, on="key", **kwargs)
```
